### PR TITLE
docs(adr): 0008 redirect hook architecture (#235)

### LINKED ISSUE #235: Spike: redirect hook architecture and ADR 0008
## Goal

Decide the architecture for the bundled redirect-hook system before any implementation begins. Output is a single ADR — no production code merges from this issue.

Parent epic: #234

## Questions to Answer

1. **Bash command parser library.** `shlex` (stdlib) vs `bashlex` (third-party PyPI dep) vs a small handwritten recursive-descent parser. The bundled hook needs to handle: `&&`, `||`, `;`, subshells `(...)`, command substitution `$(...)`, variable expansion (don't *evaluate* — just recognize that `git diff $X..HEAD` is still a `git diff`), and heredocs (recognize and skip). What's the smallest dependency footprint that gets us reliable detection?

2. **`--scope` semantics.** Mirror `claude mcp add --scope user|project|local`? Where exactly does each one write? What about precedence when entries exist in multiple scopes? What about the cases where the directory doesn't exist yet (e.g., `<repo>/.claude/` not present)?

3. **Hook script idempotency.** The hook installer must merge into existing `settings.json` without duplicating entries on re-install. What's the merge key — exact-match on `command` field? A sentinel like `"id": "git-prism-redirect"` injected into each entry? How do we handle a user who edited the installed entry and re-runs install?

4. **BDD testability.** Hook scripts run inside the Claude Code harness, not inside `git-prism` itself. How do we test the hook script from `behave`? Options: shell out to the bash script with mock JSON on stdin and assert on exit code + stderr; use a testing matrix that exercises the Python tokenizer directly; some hybrid. The tests must be hermetic and fast.

5. **Subagent MCP scope bug verification.** Issues `anthropics/claude-code#13605` and `anthropics/claude-code#13898` allegedly cause subagents to fail to call MCP tools from project-scoped `.mcp.json`. Confirm whether these issues exist, whether the bug applies to git-prism users at project scope, and whether the workaround (user-scoped registration) should be the default recommendation in the ADR.

## Definition of Done

- [ ] Spike branch `spike/redirect-hook-architecture` created and used (never merged to main)
- [ ] ADR file `docs/decisions/0008-redirect-hook-architecture.md` committed to main, structured per the format of `docs/decisions/0007-pr-125-squash-merge-post-mortem.md` (Status / Date / Context / Decision / Consequences / Prevention / Alternatives Considered / References)
- [ ] ADR records the chosen parser, the `--scope` semantics, the idempotency strategy, the BDD test approach, and the subagent-MCP-bug verification result
- [ ] Spike branch deleted after the ADR lands
- [ ] This issue is closed and linked to the ADR commit

## Out of Scope

- Production code in main (only the ADR lands)
- Implementation of the parser, hook script, or installer (that's W3 and W4)
- BDD scenarios (that's W1)

## Notes

The ADR drives the rest of the epic. W1 (BDD bootstrap) reads the ADR before writing scenarios. W3/W4 implement what the ADR specifies. If a decision in the ADR turns out to be wrong during implementation, amend the ADR — don't just diverge silently.

### PR DESCRIPTION
**Pull request by mikelane**
## Summary
- ADR 0008 records the architectural decisions for epic #234's redirect hook stack
- Five questions answered: bash parser library (`shlex` stdlib, no third-party deps), `--scope` semantics (mirror `claude mcp add`), idempotency (`id` sentinel field), BDD testability (subprocess shell-out, hermetic via tempdir), and subagent MCP scope bug verification (issue #13898 still open — defaults `--scope user` to avoid silently breaking subagents)
- 9 alternatives considered with concrete rejection reasons
- Closes #235

## SDLC gates
- Spec review (adversarial-qa): APPROVED after one revision cycle
- Bug hunting / fact-check: gh api outputs verified, issue numbers correct, REPL transcripts match shlex behavior
- Security audit: no secrets, all matches false-positive on the word "token" in technical prose
- Commit hygiene: conventional format, scoped to issue #235

## Spike discipline
Spike branch \`spike/redirect-hook-architecture\` is disposable; only the ADR file lands. Prototype code (untracked \`prototypes/\` directory in the worktree) is discarded with the branch.

🤖 Generated with [Claude Code](https://claude.ai/claude-code)

### CODE REVIEW

File: docs/decisions/0008-redirect-hook-spike.md [hunk 1]
```diff
@@ -0,0 +1,258 @@
+# ADR 0008: Redirect Hook Architecture
```
Comment (mikelane) [inline docs/decisions/0008-redirect-hook-spike.md:1]: **[NAMING — CRITICAL]** Filename does not match the established `-spike` suffix used by sibling spike ADRs in this directory (0003/0004/0005/0006 are all `-spike`). The branch is `spike/redirect-hook-architecture`, the issue (#235) DoD frames this as a spike, and `CLAUDE.md` explicitly states "The spike lives on a `spike/<topic>` branch that is never merged. Its only deliverable is an ADR in `docs/decisions/NNNN-short-title.md`."

Recommendation: rename to `0008-redirect-hook-spike.md` before merge so future readers can grep `-spike` to see all exploratory work. This is the lone blocker on the PR.

Note: 0001/0002/0007 are not `-spike` either, but those are working-tree-diff/mutation-baseline/post-mortem — not exploratory feasibility studies. The 0003-0006 cluster is the relevant precedent.

File: docs/decisions/0008-redirect-hook-architecture.md [hunk 2]
```diff
@@ -0,0 +1,258 @@
...
+| `gh pr diff 123` | hard block (exit 2) |
+| `mcp__github__get_commit` tool input | hard block |
+| Empty stdin | exit 0, no output |
+| Malformed JSON | exit 0 (fail open), warning on stderr |
+
+**Why not unit-test the Python tokenizer in isolation:**
+
+We will, for the parser logic specifically (Pythonic unit tests under `hooks/tests/test_parse_git_invocations.py`, run by `pytest`). But the behavioral contract — "this exit code, this stderr, given this stdin JSON" — is what Claude Code actually exercises, and that contract is the load-bearing surface. The BDD layer tests it end-to-end. Unit tests catch regressions in the parser internals; BDD catches regressions in the wire protocol. We need both.
+
+### 5. Subagent MCP scope bug: real, relevant, default to `--scope user`
```
Comment (mikelane) [inline docs/decisions/0008-redirect-hook-architecture.md]: **[NAMING — MEDIUM]** Heading 5 reads as comma-separated assertions ("real, relevant, default to `--scope user`") rather than a noun phrase like the other six decision headings (e.g., "Bash command parser: ...", "Idempotency: sentinel field ...", "Fail-open when `python3` is missing"). Consider rewording to a noun phrase such as "Subagent MCP scope bug workaround: default to `--scope user`" so the heading scans as a decision name rather than a sentence fragment. Deferrable to a follow-up commit on this branch.

File: docs/decisions/0008-redirect-hook-architecture.md [hunk 3]
```diff
@@ -0,0 +1,258 @@
...
+
+**Missing target directory:**
+
+If `<repo>/.claude/` does not exist for `--scope project` or `--scope local`, the installer creates it (`mkdir -p`). For `--scope user`, `~/.claude/` always exists if Claude Code is running. We never create `~/.claude/` ourselves — if it is missing, we error with "Claude Code does not appear to be installed".
+
+**Discovery (`install-hooks --status`):**
+
+The installer's `--status` flag (no positional args) reads all three settings files and prints a table of where git-prism entries are installed and which version of the hook script each one points at. This makes "is it installed?" answerable without `cat ~/.claude/settings.json | jq`.
+
+### 3. Idempotency: sentinel field `id: "git-prism-redirect-vN"`
```
Comment (mikelane) [inline docs/decisions/0008-redirect-hook-architecture.md]: **[NAMING — MEDIUM]** The sentinel id `git-prism-redirect-vN` is generic. Future schema changes might introduce a separate redirect for MCP tool inputs (already implied by the `gh pr diff` row in the coverage matrix), or for `Edit`/`Write` tools — at which point "redirect" alone disambiguates nothing. Consider `git-prism-bash-redirect-vN` to anchor the matcher dimension in the id itself. The version suffix (v1) is fine as-is. Deferrable; renaming the sentinel before any user has v1 installed is cheap.

File: docs/decisions/0008-redirect-hook-spike.md [hunk 4]
```diff
@@ -0,0 +1,258 @@
...
+
+**Hermetic constraints:**
+
+- Tests must NOT mutate `~/.claude/settings.json`. Use a `tempfile.TemporaryDirectory()` per scenario, set `HOME` to it, run the install/uninstall commands inside that sandbox, then assert on the resulting tempdir contents. Tear down by letting the tempdir context manager exit.
+- Tests must NOT spawn a real Claude Code process. The hook script reads stdin and exits — that's the entire contract. The behave tests exercise that contract directly.
+- Test fixtures live under `bdd/fixtures/hook_inputs/` as JSON files, named after the scenario they support (e.g., `bash_compound_diff.json`, `bash_status_only.json`).
+
+**Coverage matrix the BDD scenarios must cover:**
+
+| Input shape | Expected behavior |
```
Comment (mikelane) [inline docs/decisions/0008-redirect-hook-spike.md:149]: **[TEST — MEDIUM]** Coverage matrix is missing rows for several shapes the ADR itself flags as critical surface:

- command substitution `$(...)` (called out in Context as a missed case in the regex hook)
- `||` (or-list)
- `;` (sequential)
- backticks (Decision 1.2 explicitly normalizes them)
- fail-open path: `python3` missing on PATH (Decision 6 promises exit 0 + stderr warning)
- `--force` overwrite path (Decision 3)
- `--status` discovery output (Decision 2)
- `--uninstall --version v1` rollback (Decision 3)
- downgrade-version idempotency / abort path (Decision 3, "Mixed-version installs")

Each of these is a discrete behavior the BDD layer is the load-bearing test for, per Decision 4 itself. Deferrable to BDD bootstrap (#236), but the matrix in the ADR should at minimum enumerate them so #236 has a complete checklist.

File: docs/decisions/0008-redirect-hook-architecture.md [hunk 5]
```diff
@@ -0,0 +1,258 @@
...
+| `git status` | no advisory |
+| `git add file.txt` | no advisory |
+| `gh pr diff 123` | hard block (exit 2) |
+| `mcp__github__get_commit` tool input | hard block |
+| Empty stdin | exit 0, no output |
+| Malformed JSON | exit 0 (fail open), warning on stderr |
+
+**Why not unit-test the Python tokenizer in isolation:**
+
+We will, for the parser logic specifically (Pythonic unit tests under `hooks/tests/test_parse_git_invocations.py`, run by `pytest`). But the behavioral contract — "this exit code, this stderr, given this stdin JSON" — is what Claude Code actually exercises, and that contract is the load-bearing surface. The BDD layer tests it end-to-end. Unit tests catch regressions in the parser internals; BDD catches regressions in the wire protocol. We need both.
```
Comment (mikelane) [inline docs/decisions/0008-redirect-hook-architecture.md]: **[TEST — MEDIUM]** Decision 1.1 acknowledges that `shlex` newline handling is the load-bearing trick for heredoc termination, but the parser unit-test matrix is not enumerated in the ADR. At minimum the cases that should be called out for `pytest` coverage in `hooks/tests/test_parse_git_invocations.py`:

- `<<-EOF` (tab-stripping form)
- `<<'EOF'` (quoted heredoc — no expansion, but still terminated by `EOF`)
- multiple heredocs in one command (`cmd <<A <<B`)
- `EOF` token appearing inline in the body without surrounding newlines (must NOT terminate)
- malformed input: missing closing `EOF` until end-of-stream (must not crash)

Decision 4 says "Unit tests catch regressions in the parser internals" — listing the specific regressions makes that promise enforceable. Deferrable to #236.

File: docs/decisions/0008-redirect-hook-spike.md [hunk 6]
```diff
@@ -0,0 +1,258 @@
...
+- **Two wrapper functions added to the parser.** Heredoc skipping and backtick normalization. Both are <30 lines, both are unit-testable in isolation. The complexity budget for "things `shlex` doesn't handle" is bounded.
+- **One new CLI subcommand.** `git-prism install-hooks` with flags `--scope`, `--uninstall`, `--force`, `--status`. Wired in `src/main.rs` like `serve` / `manifest`.
+- **Hook entries gain a stable `id`.** Future schema changes can migrate cleanly. The version suffix is part of the contract.
+- **`--scope user` is the documented default.** The README and `--help` text both explain why. When upstream #13898 closes, this default becomes a candidate for revisit.
+- **BDD tests use subprocess shell-out, not in-process import.** Step definitions stay in Python (cross-language to Rust production code, per project BDD policy). Hermeticity comes from `tempfile`-scoped HOME, not from mocking the script's contents.
+- **No third-party Python parser dependency.** The hook works on a stock `python3` install.
+
+## Prevention
+
```
Comment (mikelane) [inline docs/decisions/0008-redirect-hook-spike.md:222]: **[TEST — MEDIUM]** No mutation-score threshold specified for the new Python tokenizer module. ADR 0002 establishes mutation testing as a first-class signal for this repo, and parsers are the highest-yield mutation target — boundary-condition mutations (off-by-one in heredoc termination, swapped `&&`/`||` split, dropped `\\n` token) are exactly what mutation testing catches that line-coverage misses.

Recommendation: set a concrete threshold (e.g., "≥85% mutation score on `parse_git_invocations.py` enforced in CI") in the Prevention section. Without one, the parser will inevitably ship with surviving mutants that escape PR review. Deferrable to #236 / W3 tokenizer issue.

File: docs/decisions/0008-redirect-hook-spike.md [hunk 7]
```diff
@@ -0,0 +1,258 @@
...
+- The `id` sentinel must be enforced in the writer and tested in BDD: a re-install scenario with an existing entry must produce no duplicates. If the sentinel is dropped or renamed, idempotency breaks silently — the BDD scenario catches it.
```
Comment (mikelane) [inline docs/decisions/0008-redirect-hook-spike.md:223]: **[TEST — MEDIUM]** Idempotency prevention claim ("a re-install scenario with an existing entry must produce no duplicates") covers only one of the four algorithm paths from Decision 3:

1. existing entry with matching id + matching path → no-op (the one currently asserted)
2. existing entry with matching id + different path → replace + log "updated"
3. existing entry with matching id but hand-edited fields → skip with warning (or `--force` overwrite)
4. existing higher-version entry (v2) when this binary writes v1 → abort with downgrade error

Each is a distinct branch in the writer logic and each can regress independently. Recommendation: expand the Prevention bullet to enumerate all four paths and tie each to a specific BDD scenario id. Deferrable to #236.

File: docs/decisions/0008-redirect-hook-spike.md [hunk 8]
```diff
@@ -0,0 +1,258 @@
...
+- stderr text: must contain a redirect message
+- stdout JSON: when emitted (advisory mode), must validate against the Claude Code hookSpecificOutput schema
+
+**Three exit-code states (matches the existing `~/.claude/hooks/git-prism-redirect.sh` prototype):**
+
+| Exit | Stdout | Stderr | Meaning |
+|---|---|---|---|
+| `0` | empty | empty | allow — command runs, hook says nothing |
+| `0` | JSON with `hookSpecificOutput.additionalContext` | optional | advise — Claude Code surfaces the message but the command still runs |
+| `2` | empty | redirect message | hard block — command is rejected, stderr is shown to the agent |
```
Comment (mikelane) [inline docs/decisions/0008-redirect-hook-spike.md:137]: **[NAMING — LOW]** "advise" (Decision 4 table, "advise" row) and "advisory" (Decision 4 prose, coverage matrix column) are used interchangeably for the middle exit-code state. Pick one register and use it consistently — either both verbs ("allow / advise / block") or both noun-phrases ("allow / advisory / hard block"). The current mix reads as inconsistent register. Trivial follow-up commit.

File: docs/decisions/0008-redirect-hook-architecture.md [hunk 9]
```diff
@@ -0,0 +1,258 @@
...
+Agents reach for raw `git diff/log/show/blame` instead of git-prism MCP tools because of pretraining bias. The current local hook (`~/.claude/hooks/git-prism-redirect.sh`) backstops this with regex pattern matching against `tool_input.command`, but the regex approach misses common shapes:
+
+- compound commands: `cd /tmp && git diff main..HEAD`
+- subshells: `(git log main..HEAD)`
+- variable expansion: `git diff $BASE..HEAD`
+- pipelines: `git diff main..HEAD | grep foo`
+- heredocs that quote git commands inside the body
+- command substitution: `cd $(git rev-parse --show-toplevel) && git diff ...`
+
+The hook also lives in user dotfiles, so installing git-prism gets the binary but none of the redirect leverage. Epic #234 ships a `git-prism install-hooks` subcommand and a bundled hook that calls a stdlib-only Python tokenizer to parse bash structurally.
```
Comment (mikelane) [inline docs/decisions/0008-redirect-hook-architecture.md]: **[NAMING — LOW]** `install-hooks` is the first hyphenated subcommand on `git-prism`. The existing subcommands listed in CLAUDE.md (`serve`, `manifest`, `snapshot`, `history`, `context`, `languages`) are all single words. As the command grows (the ADR already implies `--status`, `--uninstall`, `--force` flags), a noun-first grouping would scale better:

```
git-prism hooks install [--scope ...]
git-prism hooks uninstall [--version v1]
git-prism hooks status
```

This also matches the parallel design with `claude mcp add` / `claude mcp list` cited in Decision 2. Deferrable, but cheaper to decide before any user has `install-hooks` in muscle memory than after.

File: docs/decisions/0008-redirect-hook-architecture.md [hunk 10]
```diff
@@ -0,0 +1,258 @@
...
+
+**Decision:**
+
+- **Default `git-prism install-hooks --scope user`.** Document the reason in the CLI help text: "User scope is the default because Claude Code issue anthropics/claude-code#13898 prevents custom subagents from calling project-scoped MCP servers correctly."
+- **Document the same caveat in the README install-hooks section**, with a link to the upstream issue.
+- **Do not block `--scope project` or `--scope local`** — power users may have their own reasons (e.g., team-shared config in a monorepo, no subagent usage). Just don't make it the easy path.
+- **Re-evaluate when #13898 closes.** Add a TODO comment in `install-hooks` source that links the issue. When it closes, revisit whether `--scope project` should become the default for the install-hooks command (consistent with how MCP server registration in `.mcp.json` works).
```
Comment (mikelane) [inline docs/decisions/0008-redirect-hook-architecture.md]: **[TEST — LOW]** The decision to default `--scope user` is anchored on issue #13898 still being open. There is no re-verification mechanism — no script, no CI check, no scheduled re-poll. When #13898 closes, the trigger to revisit is purely "someone remembers."

Recommendation: a small CI job (or a scheduled `claude /loop` task) that runs `gh api repos/anthropics/claude-code/issues/13898 --jq .state` and opens a follow-up issue if the state ever flips to `closed`. Cheap to add, eliminates the "ADR drifts from reality" failure mode. Deferrable to a follow-up issue.

File: docs/decisions/0008-redirect-hook-architecture.md [hunk 11]
```diff
@@ -0,0 +1,258 @@
...
+- **No third-party Python parser dependency.** The hook works on a stock `python3` install.
+
+## Prevention
+
+- The `id` sentinel must be enforced in the writer and tested in BDD: a re-install scenario with an existing entry must produce no duplicates. If the sentinel is dropped or renamed, idempotency breaks silently — the BDD scenario catches it.
+- The shlex-edge-case list (heredoc body, backticks) must have explicit BDD scenarios, not just unit tests. If someone "improves" the parser later and accidentally regresses one of these, the BDD layer flags it.
+- The `--scope user` default must be documented in three places (CLI `--help`, README, this ADR). Drift on the default risks reverting a deliberate decision. When #13898 closes, the revisit must update all three together.
+- `parse_git_invocations.py` must be importable in tests as `from hooks.parse_git_invocations import tokenize_command`. The hook shell script must invoke it via `python3 -m parse_git_invocations` so the test path and the runtime path use the same module-loading pattern.
```
Comment (mikelane) [inline docs/decisions/0008-redirect-hook-architecture.md]: **[NAMING — LOW]** `parse_git_invocations.py` describes only the parser half. Decision 1 explicitly says the module also walks the token list and applies redirect-pattern matchers ("starts with `git diff` and contains `..`") — that is the decision half, not the parse half.

Two cleaner options:

1. Split the module: `tokenize_command.py` (pure parser) + `redirect_decision.py` (pattern matchers). Better separation of concerns, each unit-testable in isolation.
2. Rename the single module to something that names the whole job: `redirect_command_check.py` or `git_redirect_decision.py`.

Option 1 is the more idiomatic split given the mutation-testing argument in the other comment — pure parsers and decision tables mutate very differently. Deferrable.

Comment (mikelane) [inline docs/decisions/0008-redirect-hook-architecture.md]: **[PYTHON — LOW]** The Prevention section asserts two import patterns:

- tests: `from hooks.parse_git_invocations import tokenize_command`
- runtime: `python3 -m parse_git_invocations`

These two patterns require different `sys.path` shapes:

- The first requires the parent of `hooks/` to be on `sys.path` (so `hooks` is an importable package — needs `hooks/__init__.py`).
- The second requires `hooks/` itself to be on `sys.path` (so `parse_git_invocations` is a top-level module reachable via `-m`).

Reconciling these requires either: (a) `pytest` setting `pythonpath = ["."]` in `pyproject.toml` AND the hook script doing `cd "$(dirname "$0")" && python3 -m parse_git_invocations`, or (b) a `conftest.py` that adjusts `sys.path` for tests. The ADR does not commit to one. Recommend stating the resolution explicitly so the W3 implementer does not have to re-derive it. Deferrable to W3.

**Review by mikelane — COMMENTED** [review]
## Review Summary

This is a strong spike ADR. The decisions are evidence-anchored (live REPL transcripts for `shlex` behavior, pasted `gh api` responses for the upstream MCP scope bug verification), and the alternatives section earns its keep — nine rejected options with concrete reasoning is the right amount of due diligence for a load-bearing architectural choice. The "fail-open when `python3` is missing" decision and the "`--scope user` because of upstream issue #13898" decision are both well-argued and would survive adversarial review on their own merits.

### Decision

**CHANGES REQUESTED** — single blocker on filename naming.

### Blockers (must fix before merge)

1. **CRITICAL: Filename does not match the established `-spike` suffix** for spike ADRs in this directory (0003/0004/0005/0006 all carry it; 0001/0002/0007 don't, but those aren't feasibility spikes). The branch is `spike/redirect-hook-architecture`, the issue (#235) is framed as a spike, and `CLAUDE.md` is explicit: "The spike lives on a `spike/<topic>` branch that is never merged. Its only deliverable is an ADR in `docs/decisions/NNNN-short-title.md`." Rename to `0008-redirect-hook-spike.md` before merge so the next reader can grep `-spike` and see all exploratory work.

### Recommendations (deferrable, do not block merge)

The remaining 11 inline findings split cleanly across three buckets:

- **Defer to BDD bootstrap (#236):** coverage matrix completeness (command substitution, `||`, `;`, backticks, fail-open, `--force`, `--status`, `--uninstall`, downgrade), heredoc parser test matrix (`<<-EOF`, quoted heredoc, multiple, malformed), mutation-score threshold for the new tokenizer, four-path idempotency prevention.
- **Defer to W3 / W4 implementation issues:** module split or rename for `parse_git_invocations.py`, `sys.path` reconciliation between `from hooks.parse_git_invocations import ...` (tests) and `python3 -m parse_git_invocations` (runtime).
- **Trivial follow-up commit on this branch (or post-merge fixup):** Decision 5 heading wording (sentence fragment vs. noun phrase), sentinel id specificity (`git-prism-bash-redirect-vN`), `advise`/`advisory` register, `install-hooks` vs. `git-prism hooks install` subcommand grouping, upstream-issue re-verification cadence.

### Praise

- The Python 3.13 REPL transcript inline in Decision 1 is exemplary — readers can reproduce the evidence in seconds rather than taking the ADR's word for it.
- The "Why a sentinel `id` field over command-string match" sub-decision in Decision 3 (and the explicit rejection of the comment-marker alternative) is the level of "show your work" that makes ADRs durable.
- The `claude mcp add` parallel design rationale in Decision 2 — and the deliberate divergence on the default scope — is well-justified, with both reasons (subagent MCP bug AND user-scope-applies-everywhere) explicitly enumerated.
- The "Mixed-version installs (downgrade after upgrade)" subsection of Decision 3 is the kind of edge case authors usually omit and field reports surface six months later. Catching it in the spike is the right call.
- The Alternatives Considered section, particularly the rejection of `bashlex` (license + bootstrap cost) and tree-sitter (process boundary), is thorough enough that future revisitors won't have to re-litigate.

### Security

Zero findings. No code lands; future implementation issues (#237, #239) get their own gauntlet.

### Solo-dev note

Per `CLAUDE.md`, GitHub doesn't allow repo owners to formally approve their own PRs, so this is posted as a `COMMENT` review rather than `APPROVE`. The filename rename is the only blocker — once `0008-redirect-hook-architecture.md` is renamed to `0008-redirect-hook-spike.md`, the PR can be merged.

Generated with [Claude Code](https://claude.ai/claude-code)
